src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/AttachCaseCallbackService.java

```diff
@@ -42,67 +51,153 @@ public AttachCaseCallbackService(CcdApi ccdApi) {
     /**
      * Attaches exception record to a case.
      *
-     * @return List of errors
+     * @return Either a map of fields the should be modified in CCD, when processing was successful,
+     *         or the list of errors, in case of errors
      */
-    public List<String> process(CaseDetails exceptionRecord) {
+    public Either<List<String>, Map<String, Object>> process(CaseDetails exceptionRecord) {
+        boolean useSearchCaseReference = isSearchCaseReferenceTypePresent(exceptionRecord);
+
+        return getValidation(exceptionRecord, useSearchCaseReference)
+            .map(this::tryAttachToCase)
+            .getOrElseGet(errors -> Either.left(errors.toJavaList()));
+    }
+
+    private Validation<Seq<String>, AttachToCaseEventData> getValidation(
+        CaseDetails exceptionRecord,
+        boolean useSearchCaseReference
+    ) {
+        Validation<String, String> caseReferenceTypeValidation = useSearchCaseReference
+            ? hasSearchCaseReferenceType(exceptionRecord)
+            : valid(CCD_CASE_REFERENCE);
+
+        Validation<String, String> caseReferenceValidation = useSearchCaseReference
+            ? hasSearchCaseReference(exceptionRecord)
+            : hasAttachToCaseReference(exceptionRecord);
+
         return Validation
             .combine(
                 hasJurisdiction(exceptionRecord),
-                hasAttachToCaseReference(exceptionRecord),
+                hasServiceNameInCaseTypeId(exceptionRecord),
+                caseReferenceTypeValidation,
+                caseReferenceValidation,
                 hasAnId(exceptionRecord),
                 hasAScannedRecord(exceptionRecord)
             )
-            .ap(this::attachToCase)
-            .getOrElseGet(Value::toJavaList);
+            .ap(AttachToCaseEventData::new);
     }
 
     //The code below need to be rewritten to reuse the EventPublisher class
 
-    private List<String> attachToCase(
-        String exceptionRecordJurisdiction,
-        String targetCaseRef,
-        Long exceptionRecordId,
-        List<Map<String, Object>> exceptionRecordDocuments
+    private Either<List<String>, Map<String, Object>> tryAttachToCase(
```

Comment:
comment explaining what's in the 'either' object would be nice

Reply:
I've got it on the main method. Can add it here, too.